include/deal.II/base/index_set.h

```diff
@@ -79,6 +89,188 @@ class IndexSet
   typedef signed int value_type;
 
   /**
+   * Dereferencing an IntervalIterator will return a reference to an object of this type.
+   */
+  class IntervalAccessor
+  {
+  public:
+    IntervalAccessor(const IndexSet &idxset, size_type range_idx);
+    size_type n_elements() const;
+    bool is_valid() const;
+    ElementIterator begin() const;
+    ElementIterator end() const;
+  private:
+    bool operator == (const IntervalAccessor &other) const;
+    bool operator < (const IntervalAccessor &other) const;
+    void advance ();
+    const IndexSet &index_set;
+    size_type range_idx;
+
+    friend class IntervalIterator;
+  };
+
+  /**
+   * Class that represents an iterator pointing to a contiguous interval [a,b[ as returned by IndexSet::begin_interval().
+   */
+  class IntervalIterator
+  {
+  public:
+    /**
+     * Construct a valid iterator pointing to the interval with index @p range_idx.
+     */
+    IntervalIterator(const IndexSet &idxset, size_type range_idx);
+
+    /**
+     * Construct an invalid iterator (used as end())
+     */
+    IntervalIterator(const IndexSet &idxset);
+
+    /**
+     * Prefix increment.
+     */
+    IntervalIterator &operator++ ();
+
+    /**
+     * Postfix increment.
+     */
+    IntervalIterator operator++ (int);
+
+    /**
+     * Dereferencing operator, returns an IntervalAccessor.
+     */
+    const IntervalAccessor &operator* () const;
```

Comment:
Any particular reason to return a reference? What happens in 
```
  auto p = *iterator++;
```
Does `p` now reference the old or new interval?